File: scripts/trial/actions/count_until_client.py
#!/usr/bin/env python
import rospy
import actionlib

from agv_motion.msg import CountUntilAction
from agv_motion.msg import CountUntilGoal

class CountUntilClient:
    def __init__(self):
        self._ac = actionlib.SimpleActionClient(
            "/count_until", 
            CountUntilAction)
        self._ac.wait_for_server()  # wait for server to be up (before sending a goal)
        rospy.loginfo("Action server is up, client can send new goals")

    def send_goal_and_get_result(self):
        goal = CountUntilGoal(max_number=15, wait_duration=0.75)
        self._ac.send_goal(goal, done_cb=self.done_callback, feedback_cb=self.feedback_callback)
        # done_cb is triggered when server sets goal to terminal state - succeeded
        rospy.loginfo("Goal has been sent")

        rospy.sleep(2)
        self._ac.cancel_goal()
        
        ## syncronous
        ## block until sever sends the results
        ## if it takes more than duration, then stop waiting and exit
        # success = self._ac.wait_for_result(rospy.Duration(30.0))  
        # if not success: 
        #     rospy.logwarn("Timeout reached")

        # rospy.loginfo(self._ac.get_result())

    def done_callback(self, status, result):
        # http://docs.ros.org/en/diamondback/api/actionlib_msgs/html/msg/GoalStatus.html
        rospy.loginfo("Status is : " + str(status)) 
        rospy.loginfo("Result is : " + str(result))

    def feedback_callback(self, feedback):
        rospy.loginfo(feedback)


if __name__ == '__main__':
    rospy.init_node('count_until_client')

    client = CountUntilClient()

    client.send_goal_and_get_result()

    rospy.spin() # needs to keep on runnning in asynchronous mode (when done_cb is used)

File: src/trial/actions/count_until_client.cpp
#include <ros/ros.h>
#include <actionlib/client/simple_action_client.h>
#include <actionlib/client/terminal_state.h>

#include <agv_motion/CountUntilAction.h>
#include <agv_motion/CountUntilGoal.h>
#include <agv_motion/CountUntilResult.h>
#include <agv_motion/CountUntilFeedback.h>

class CountUntilClient
{
protected:
    ros::NodeHandle _nh;
    actionlib::SimpleActionClient<agv_motion::CountUntilAction> _ac;

public:
    CountUntilClient() : _ac("/count_until", true)
    {
        ROS_INFO("Wait for the action server to start...");

        _ac.waitForServer();

        ROS_INFO("Server is up.");
    }

    void sendGoal()
    {
        agv_motion::CountUntilGoal goal;
        goal.max_number = 12;
        goal.wait_duration = 0.3;
        _ac.sendGoal(
            goal,
            boost::bind(&CountUntilClient::doneCb, this, _1, _2), // necessary when using callbacks in class
            boost::bind(&CountUntilClient::activeCb, this),
            boost::bind(&CountUntilClient::feedbackCb, this, _1));
        ROS_INFO("Goal has been sent.");

        // Cancel the goal after 2 seconds
        // ros::Duration(2.0).sleep();
        // _ac.cancelGoal();
    }

    // when the goal reached terminal state (preempted, succeeded, aborted)
    void doneCb(const actionlib::SimpleClientGoalState &state,
                const agv_motion::CountUntilResultConstPtr &result)
    {
        ROS_INFO("Finished in state: %s", state.toString().c_str());
        ROS_INFO("Count result: %d", (int)result->count);
    }

    // when the goal went active
    void activeCb()
    {
        ROS_INFO("Goal just went active");
    }

    void feedbackCb(const agv_motion::CountUntilFeedbackConstPtr &feedback)
    {
        ROS_INFO("Feedback received. Percentage: %lf", feedback->percentage);
    }
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "count_until_client");

    ROS_INFO("Starting client...");

    CountUntilClient client;

    client.sendGoal(); // send goal using client

    ros::spin();
}